core/src/test/java/com/linecorp/armeria/server/InitiateConnectionShutdownTest.java

```diff
@@ -0,0 +1,350 @@
+    @CsvSource({
+            "/goaway_async?duration=1",
+            "/goaway_blocking?duration=1",
+    })
+    void initiateConnectionShutdownWithDrainHttp2(String path) throws Exception {
+        makeHttp2Request(path);
+        final InOrder inOrder = inOrder(clientListener);
+        inOrder.verify(clientListener).onGoAwayRead(any(ChannelHandlerContext.class), eq(Integer.MAX_VALUE),
+                                                    eq(Http2Error.NO_ERROR.code()), eq(DEBUG_DATA));
+        inOrder.verify(clientListener).onGoAwayRead(any(ChannelHandlerContext.class), eq(STREAM_ID),
+                                                    eq(Http2Error.NO_ERROR.code()), eq(DEBUG_DATA));
+        inOrder.verify(clientListener).onDataRead(any(ChannelHandlerContext.class), eq(STREAM_ID),
+                                                  any(ByteBuf.class), anyInt(), eq(true));
+    }
+
+    @ParameterizedTest
+    @CsvSource({
+            "/goaway_async?duration=500",
+            "/goaway_blocking?duration=500",
+    })
+    void initiateConnectionShutdownCloseBeforeDrainEndHttp2(String path) throws Exception {
+        makeHttp2Request(path);
+        final InOrder inOrder = inOrder(clientListener);
+        inOrder.verify(clientListener).onGoAwayRead(any(ChannelHandlerContext.class), eq(Integer.MAX_VALUE),
+                                                    eq(Http2Error.NO_ERROR.code()), eq(DEBUG_DATA));
+        inOrder.verify(clientListener).onDataRead(any(ChannelHandlerContext.class), eq(STREAM_ID),
+                                                  any(ByteBuf.class), anyInt(), eq(true));
+        inOrder.verify(clientListener).onGoAwayRead(any(ChannelHandlerContext.class), eq(STREAM_ID),
+                                                    eq(Http2Error.NO_ERROR.code()), eq(DEBUG_DATA));
+    }
+
+    @ParameterizedTest
+    @CsvSource({
+            "/goaway_async",
+            "/goaway_async?duration=-1",
+            "/goaway_async?duration=0",
+            "/goaway_async?duration=1",
+            "/goaway_async?duration=500",
+            "/goaway_blocking",
+            "/goaway_blocking?duration=-1",
+            "/goaway_blocking?duration=0",
+            "/goaway_async?duration=1",
+            "/goaway_async?duration=500",
+    })
+    void initiateConnectionShutdownHttp1(String path) throws Exception {
+        try (CloseableHttpClient hc = HttpClients.createMinimal()) {
+            final HttpUriRequest req = new HttpGet(goAwayServer.httpUri() + path);
+            try (CloseableHttpResponse res = hc.execute(req)) {
+                assertThat(res.getStatusLine().toString()).isEqualTo("HTTP/1.1 200 OK");
+                assertThat(res.containsHeader("Connection")).isTrue();
+                assertThat(res.getHeaders("Connection"))
+                        .extracting(Header::getValue).containsExactly("close");
+            }
+        }
+        await().timeout(Duration.ofSeconds(2)).untilTrue(connectionClosed);
```

Comment:
The default timeout (10 seconds) sounds a little bit long, but wouldn't it help reducing flakiness?

Reply:
SG, I was just annoyed with long awaits while debugging local issues.